### PR TITLE
Public API, part 1

### PR DESCRIPTION
**Pull request by JohnLCaron**
1) Make the core objects (ucar.nc2.*, ucar.nc2.datatset.*) immutable by adding builders.
2) Mark mutating methods deprecated.
3) Start to move static helpers to seperate class.
4) Move/port non-public classes to ucar.nc2.internal.

This should not be submitted until 5.2 branch is released.
In any case theres a lot more to do.

---
This change is [Reviewable](https://reviewable.io/reviews/unidata/netcdf-java/137)

### CODE REVIEW

**Comment by JohnLCaron** [comment]
Ok, this is ready as soon as 5.2 is tagged.
Would be good to do asap so I can see the diffs for the next PR (coord system) which is pretty complicated. thanks

**Comment by lesserwhirls** [comment]
Awesome! 5.2 is now cut on 5.3.0-SNAPSHOT is on master. Would you mind rebasing?

**Comment by JohnLCaron** [comment]
done

File: cdm/core/src/test/java/ucar/nc2/ft2/simpgeometry/TestSimpleGeometryFeature.java [hunk 1]
```diff
@@ -1,93 +0,0 @@
-package ucar.nc2.ft2.simpgeometry;
```
Comment (lesserwhirls) [inline cdm/core/src/test/java/ucar/nc2/ft2/simpgeometry/TestSimpleGeometryFeature.java]: Was it not possible to save this test given the changes?

Comment (JohnLCaron) [reply cdm/core/src/test/java/ucar/nc2/ft2/simpgeometry/TestSimpleGeometryFeature.java]: Not sure how/why that got removed. Now erstored, and seems to pass.

**Comment by lesserwhirls** [comment]
Looks like just a spotless issue on the restored file. Once that's done we'll be good to go on this one.

**Comment by lesserwhirls** [comment]
I went ahead and applied spotless and pull this in over on Unidata/netcdf-java#142. Thanks @JohnLCaron!

**Pull request merged**